synapse/storage/data_stores/main/events.py

```diff
@@ -1614,3 +1626,87 @@ def f(txn, stream_ordering):
             await self.db.runInteraction("locally_reject_invite", f, stream_ordering)
 
         return stream_ordering
+
+    def _maybe_insert_unread_event_txn(
+        self, txn: Connection, event: EventBase, context: EventContext,
+    ):
+        """Mark the event as unread for every current member of the room if it passes the
+        conditions for that.
+
+        These conditions are: the event must either have a non-empty string body, be an
+        encrypted message, or be either a power levels event, a room name event or a room
+        topic event, and must be neither rejected or soft-failed nor an edit or a notice.
+
+        Args:
+            txn: The transaction to use to retrieve room members and to mark the event
+                as unread.
+            event: The event to evaluate and maybe mark as unread.
+            context: The context in which the event was sent (used to figure out whether
+                the event has been rejected).
+        """
+        content = event.content
+
+        is_edit = (
+            content.get("m.relates_to", {}).get("rel_type") == RelationTypes.REPLACE
+        )
+        is_notice = not event.is_state() and content.get("msgtype") == "m.notice"
+
+        # We don't want rejected or soft-failed events, edits or notices to be marked
+        # unread.
+        if (
+            context.rejected
+            or is_edit
+            or is_notice
+            or event.internal_metadata.is_soft_failed()
+        ):
+            return
+
+        body_exists = isinstance(content.get("body"), str)
+        is_state_event_to_mark_unread = (
+            event.is_state() and event.type in STATE_EVENT_TYPES_TO_MARK_UNREAD
+        )
+        is_encrypted_message = (
+            not event.is_state() and event.type == EventTypes.Encrypted
+        )
+
+        # We want to mark unread messages with a non-empty string body, some state events
+        # (power levels, room name, room topic, room avatar) and encrypted messages.
+        if not (body_exists or is_state_event_to_mark_unread or is_encrypted_message):
+            return
+
+        # Get the list of users that are currently joined to the room.
+        users_in_room = self.db.simple_select_onecol_txn(
+            txn=txn,
+            table="current_state_events",
+            keyvalues={"membership": Membership.JOIN, "room_id": event.room_id},
+            retcol="state_key",
+        )  # type: list
+
+        # Only insert rows for local users.
+        local_users_in_room = list(
+            filter(lambda user_id: self.hs.is_mine_id(user_id), users_in_room)
+        )
+
+        # Mark the message as unread for every user currently in the room, except the
+        # sender of the event (because even if they haven't sent a read receipt for the
+        # event, it seems dumb to show it as unread to its sender).
```

Comment:
this logic seems sane, but is probably worth making explicit in the MSC.

Reply:
Done: https://github.com/matrix-org/matrix-doc/pull/2654/commits/928a8dff2efbd77064e3320d12b117a69f8169f4